// Copyright (C) 2018,2019,2021 Rodrigo Jose Hernandez Cordoba
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

syntax = "proto3";
package AeonGames;

import "transform.proto";

message ComponentPropertyMsg{
    string name = 1;
    oneof value {
    sint32    int = 2;
    sint32    long = 3;
    sint64    long_long = 4;
    uint32    unsigned = 5;
    uint32    unsigned_long = 6;
    uint64    unsigned_long_long = 7;
    float     float = 8;
    double    double = 9;
    string    string = 10;
    string    path = 11;
    }
}

message ComponentMsg {
    string name = 1;
    repeated ComponentPropertyMsg property = 2;
}

message NodeMsg {
    string name = 1;
    oneof transform {
        TransformMsg local = 2;
        TransformMsg global = 3;
    }
    repeated ComponentMsg component = 4;
    repeated NodeMsg node = 5;
}

message CameraMsg {
    string              node          = 1;
    float               field_of_view = 2;
    float               near_plane    = 3;
    float               far_plane     = 4;
}

message SceneMsg{
    string              name   = 1;
    CameraMsg        camera = 2;
    repeated NodeMsg node   = 3;
}
